### PR TITLE
Add `choose_conv_method`

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**
This PR blocks #3371.
It is an implementation of `scipy.signal.choose_conv_method ` (supporting only 1D inputs currently). One difference from SciPy's behaviour is that it always chooses direct convolution for signed/unsigned integers

### CODE REVIEW

File: cupyx/scipy/signal/signaltools.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        This function currently doesn't support measure option,
+        nor multidimensional inputs.
+
+    .. seealso:: :func:`scipy.signal.choose_conv_method`
+
+    """
+    if in1.ndim != 1 or in2.ndim != 1:
+        raise NotImplementedError('Only 1d inputs are supported currently')
+
+    if _numeric_arrays([in1, in2], kinds='bui'):
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:37]: `_numeric_arrays` is used only from here. How about embedding directly?
```suggestion
    if in1.dtype.kind in 'bui' or in2.dtype.kind in 'bui':
```

File: cupyx/scipy/signal/signaltools.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    """Find the fastest convolution/correlation method.
+
+    Args:
+        in1 (cupy.ndarray): first input.
+        in2 (cupy.ndarray): second input.
+        mode (str, optional): `valid`, `same`, `full`
+
+    Returns:
+        str: A string indicating which convolution method is fastest,
+         either ‘direct’ or ‘fft’.
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:25]: Correct indentation and quotations.
```suggestion
        either ``direct`` or ``fft1``.
```

File: cupyx/scipy/signal/signaltools.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    return True
+
+
+def choose_conv_method(in1, in2, mode='full'):
+    """Find the fastest convolution/correlation method.
+
+    Args:
+        in1 (cupy.ndarray): first input.
+        in2 (cupy.ndarray): second input.
+        mode (str, optional): `valid`, `same`, `full`
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:21]:
```suggestion
        mode (str, optional): ``valid``, ``same``, ``full``
```

File: cupyx/scipy/signal/signaltools.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def _fftconv_faster(x, h, mode):
+    """
+    .. seealso:: :func: `scipy.signal.signaltools._fftconv_faster`
+
+    """
+    fft_ops, direct_ops = _conv_ops(x.size, h.size, mode)
+    offset = -1e-3
+    constants = {
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:53]: Please explain what this constants are. I'm wondering these constants are useful also for GPU computations.

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+class TestChooseConvMethod(unittest.TestCase):
+
+    @testing.for_dtypes('efdFD')
+    @testing.numpy_cupy_equal(scipy_name='scp')
+    def test_choose_conv_method1(self, xp, scp, dtype):
+        a = testing.shaped_arange((10,), xp, dtype)
+        b = testing.shaped_arange((5,), xp, dtype)
+        return scp.signal.choose_conv_method(a, b, mode=self.mode)
+
+    @testing.for_dtypes('bBhHiIlLqQpP')
```
Comment (asi1024) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:26]:
```suggestion
    @testing.for_int_dtypes(no_bool=True)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import cupy
+import cupyx
+from cupy import testing
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'mode': ['valid', 'same', 'full'],
+    'shape': [((), (5,)), ((3, 4, 5), (1, 2))]
```
Comment (asi1024) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:14]: This parameter is used only for `test_choose_conv_method_ndim`. Please split the test class.

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+}))
+@testing.with_requires('scipy')
+class TestChooseConvMethod(unittest.TestCase):
```
Comment (asi1024) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:17]: Please add a test for `siz1 == siz2`.

File: cupyx/scipy/signal/signaltools.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    Args:
+        in1 (cupy.ndarray): first input.
+        in2 (cupy.ndarray): second input.
+        mode (str, optional): `valid`, `same`, `full`
+
+    Returns:
+        str: A string indicating which convolution method is fastest,
+         either ‘direct’ or ‘fft’.
+
+    .. warning::
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:27]: Please write the difference of the behavior from the SciPy's one in this docstring and the description of this PR.

File: cupyx/scipy/signal/signaltools.py (hunk 4 above)
Comment (grlee77) [reply cupyx/scipy/signal/signaltools.py:53]: These constants for SciPy were determined using code from https://github.com/stsievert/scipy-convolve-timings as discussed at:  https://github.com/scipy/scipy/pull/11031#issuecomment-552068555

I expect running the same sort of thing on the GPU might give substantially different results. I am not sure how feasible it is to get a single set of numbers that applies in general across a wide range of GPUs, but a single set based on a given GPU would probably still be a better match than these existing CPU-based constants.

File: cupyx/scipy/signal/signaltools.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+    Returns:
+        str: A string indicating which convolution method is fastest,
+        either ``direct`` or ``fft1``.
+
+    .. warning::
+        This function currently doesn't support measure option,
+        nor multidimensional inputs. It differs from SciPy's
+        behavior in handling signed and unsigned integers by
+        always choosing ``direct`` convolution method.
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:20]: How about writing just `This function does not guarantee the compatibility of the return value to NumPy's one.`?

File: cupyx/scipy/signal/signaltools.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    .. seealso:: :func: `scipy.signal.signaltools._fftconv_faster`
+
+    """
+    fft_ops, direct_ops = _conv_ops(x.size, h.size, mode)
+    # TODO(Dahlia-Chehata): replace with GPU-based constants.
+    return True
+
+
+def _conv_ops(siz1, siz2, mode):
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:46]: Please remove the unused function in this PR.

Comment (Dahlia-Chehata) [reply cupyx/scipy/signal/signaltools.py:46]: What about the tests ?

Comment (asi1024) [reply cupyx/scipy/signal/signaltools.py:46]: `_conv_ops` is not tested in this PR.

Comment (Dahlia-Chehata) [reply cupyx/scipy/signal/signaltools.py:46]: Yes, I know. I mean should I eliminate the tests related to dtypes ('efdFD') ?

File: cupyx/scipy/signal/signaltools.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        mode (str, optional): ``valid``, ``same``, ``full``.
+
+    Returns:
+        str: A string indicating which convolution method is fastest,
+        either ``direct`` or ``fft1``.
+
+    .. warning::
+        This function currently doesn't support measure option,
+        nor multidimensional inputs. It does not guarantee
+        the compatibility of the return value to NumPy's one.
```
Comment (asi1024) [inline cupyx/scipy/signal/signaltools.py:17]:
```suggestion
        the compatibility of the return value to SciPy's one.
```

File: cupyx/scipy/signal/signaltools.py (hunk 10 above)
Comment (asi1024) [reply cupyx/scipy/signal/signaltools.py]: OK to remove `test_choose_conv_method_same`.

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 12]
```diff
@@ -18,23 +18,17 @@ class TestChooseConvMethod(unittest.TestCase):
     @testing.for_dtypes('efdFD')
     @testing.numpy_cupy_equal(scipy_name='scp')
     def test_choose_conv_method1(self, xp, scp, dtype):
-        a = testing.shaped_arange((10,), xp, dtype)
-        b = testing.shaped_arange((5,), xp, dtype)
+        a = testing.shaped_arange((10000,), xp, dtype)
+        b = testing.shaped_arange((5000,), xp, dtype)
         return scp.signal.choose_conv_method(a, b, mode=self.mode)
```
Comment (asi1024) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:23]: `cupy.choose_conv_method` supported in this PR does not guarantee the compatibility to SciPy's one. So please test it without `numpy_cupy_equal`.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**